### PR TITLE
[c++/python] Add `tiledb_timestamp` argument for `reopen`

### PR DESCRIPTION
**Pull request by nguyenv**
**Issue and/or context:**

As reported internally, the `reopen` unit tests intermittenly fail due to a race condition when objects are opened at the same timestamp.

**Changes:**

Add `tiledb_timestamp` for `reopen` methods and modify tests to use explicit timestamps to prevent race conditions.

### CODE REVIEW

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
Suggestion:  the timestamp param for reopen also accepts a `datetime.datetime` (as does the timestamp for open, etc).  Can we add at least one test to confirm this works as expected, in addition to the existing tests which cover None and int (ms).

Otherwise it looks good to me.

**Review by bkmartinjr — APPROVED** [review]

File: apis/python/tests/test_sparse_nd_array.py [hunk 1]
```diff
@@ -84,33 +85,38 @@ def test_sparse_nd_array_reopen(tmp_path):
...
+            with A2.reopen("r", tiledb_timestamp=3) as A3:
                 assert A1.mode == "r"
                 assert A2.mode == "w"
                 assert A3.mode == "r"
-                assert A1.tiledb_timestamp < A2.tiledb_timestamp
-                assert A2.tiledb_timestamp < A3.tiledb_timestamp
-
-    with soma.SparseNDArray.open(tmp_path.as_posix(), "r") as A1:
-        with A1.reopen("r") as A2:
+                assert A1.tiledb_timestamp.timestamp() == 0.001
```
Comment (johnkerl) [inline apis/python/tests/test_sparse_nd_array.py:97]: I'd prefer `A1.tiledb_timestamp_ms` so we can do int-to-int compares rather than relying on whether floating-point equality rounds correctly (there's no need for floating-point here)

**Review by johnkerl — CHANGES_REQUESTED** [review]

**Review by johnkerl — APPROVED** [review]

**Review by johnkerl — APPROVED** [review]
